src/v/archival/ntp_archiver_service.cc

```diff
@@ -854,4 +858,97 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+                          // scheulding group here.
+                          return ss::with_scheduling_group(_upload_sg, [this] {
+                              return apply_retention()
+                                .then([this]() { return garbage_collect(); })
+                                .finally([this] {
+                                    if (_gate.is_closed()) {
+                                        return;
+                                    }
+                                    _housekeeping_timer.rearm(
+                                      _housekeeping_jitter());
+                                });
+                          });
+                      }).handle_exception([this](std::exception_ptr e) {
+        vlog(_rtclog.info, "Error in housekeeping timer callback: {}", e);
+    });
+}
+
+ss::future<> ntp_archiver::apply_retention() {
+    auto log = _partition_manager.log(_ntp);
+    if (!log) {
+        co_return;
+    }
+
+    auto retention_calculator = retention_calculator::factory(
+      manifest(), _partition->get_ntp_config(), *log);
+    if (!retention_calculator) {
+        co_return;
+    }
+
+    auto next_start_offset = retention_calculator->next_start_offset();
+    if (next_start_offset) {
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _partition->archival_meta_stm()->truncate(
+          *next_start_offset, deadline, _as);
+        if (error != cluster::errc::success) {
+            vlog(
+              _rtclog.warn,
+              "Failed to update archival metadata STM start offest according "
+              "to retention policy: {}",
+              error);
+        }
+    }
+}
+
+ss::future<> ntp_archiver::garbage_collect() {
+    std::vector<remote_segment_path> to_remove;
+    auto current_start_offset = manifest().get_start_offset();
+
+    for (const auto& [key, value] : manifest()) {
+        if (value.committed_offset < current_start_offset) {
+            to_remove.emplace_back(
+              manifest().generate_remote_segment_path(_ntp, key, value));
+        }
+    }
+
+    auto replaced = manifest().replaced_segments();
+    for (const auto& meta : replaced) {
+        const auto segment_name = cloud_storage::generate_local_segment_name(
+          meta.base_offset, meta.segment_term);
+        to_remove.emplace_back(cloud_storage::generate_remote_segment_path(
+          _ntp, meta.ntp_revision, segment_name, meta.archiver_term));
+    }
+
+    // TODO(vlad): Make this loop concurrent.
+    for (const auto& path : to_remove) {
+        retry_chain_node fib(
+          _metadata_sync_timeout, _cloud_storage_initial_backoff, &_rtcnode);
+        retry_chain_logger ctxlog(archival_log, fib, _ntp.path());
+
+        // If deletion fails and the subsequent command to cleanup metadata
```

Comment:
We probably want roughly the same error handling here as in remote_partition's deletion code: timeouts are transient + we should retry later, everything else should prompt us to abandoned the object in S3.